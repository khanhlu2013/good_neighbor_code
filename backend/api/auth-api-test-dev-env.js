const route = require("express").Router();

const keys = require("../configs/keys");
const { FRONTEND_URL } = keys;

if (keys.NODE_ENV === "production") {
  throw Error("Unexpect production env");
}

route.get("/dummy_login_for_test_purpose", (req, res, next) => {
  const User = require("../models/user");
  const { email, name } = req.query;

  (async () => {
    const user = await User.findOneOrCreate(email, name);
    req.login(user, err => {
      if (err) {
        console.log(err);
        res.status(500).send();
      } else {
        return res.redirect(FRONTEND_URL);
      }
    });
  })().catch(next);
});

module.exports = route;
