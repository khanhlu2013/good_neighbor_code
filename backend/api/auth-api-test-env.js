const route = require('express').Router();

const keys = require('../configs/keys');
const {FRONTEND_URL} = keys;

if(keys.NODE_ENV !== 'test'){
  throw Error(`Unexpected ${keys.NODE_ENV} env`);
}

route.get('/dummy_login_for_test_purpose',async (req,res)=>{
  const User =  require('../models/user');
  const {email,name} = req.query;
  try{
    const user = await User.findOneOrCreate(email,name);
    req.login(user,(err)=>{
      if(err){
        console.log(err);
        res.status(500).send();
      }else{
        return res.redirect(FRONTEND_URL);
      }
    });
  }catch(e){
    console.log(e);
    res.status(500).send();
  }
});

module.exports = route;