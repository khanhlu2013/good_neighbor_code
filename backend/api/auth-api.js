const route = require("express").Router();
const passport = require("passport");

const keys = require("../configs/keys");

route.get(
  "/google",
  passport.authenticate("google", { scope: ["profile", "email"] })
);
route.get("/google/redirect", passport.authenticate("google"), (req, res) => {
  res.redirect(keys.FRONTEND_URL);
});

if (keys.NODE_ENV !== "production") {
  route.get("/login_for_test_dev", (req, res, next) => {
    const User = require("../models/user");
    const { email, name } = req.query;
    (async () => {
      const user = await User.findOneOrCreate(email, name);
      req.login(user, err => {
        if (err) {
          console.log(err);
          res.status(500).send();
        } else {
          return res.redirect(keys.FRONTEND_URL);
        }
      });
    })().catch(next);
  });
}

module.exports = route;
