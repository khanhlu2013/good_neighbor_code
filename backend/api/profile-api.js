const route = require("express").Router();
const mongoose = require("mongoose");

const Connection = require("../models/connection");
const Post = require("../models/post");
const User = require("../models/user");
const ShareLog = require("../models/shareLog");
const keys = require("../configs/keys");

// - auth ----
const authCheck = (req, res, next) => {
  if (!req.user) {
    return res.status(401).send();
  }
  next();
};

route.get("/logout", authCheck, (req, res) => {
  req.logout();
  res.redirect(keys.FRONTEND_URL);
});

route.get("/", authCheck, (req, res) => {
  res.send(req.user);
});

// - user ----
route.get("/searchEmail", authCheck, (req, res, next) => {
  const { email } = req.query;
  const { user } = req;

  if (!email) {
    return res.status(400).send();
  }

  if (email === user.email) {
    return res.status(400).send();
  }

  (async () => {
    res.send(await User.findOne({ email }));
  })().catch(next);
});

// - connection ----
route.get("/connections", authCheck, (req, res, next) => {
  const { user } = req;
  (async () => {
    const connections = await Connection.find({
      $or: [{ from: user }, { to: user }]
    })
      .populate("from")
      .populate("to");
    res.send(connections);
  })().catch(next);
});

route.post("/createConnection", authCheck, (req, res, next) => {
  const { user } = req;
  const { userIdToAdd } = req.body;

  if (!userIdToAdd) {
    return res.status(400).send();
  }

  if (!mongoose.Types.ObjectId.isValid(userIdToAdd)) {
    return res.status(400).send();
  }

  (async () => {
    const connection = await new Connection({
      from: user.id,
      to: userIdToAdd
    }).save();
    return res.send(connection);
  })().catch(next);
});

route.post("/updateConnection", authCheck, (req, res, next) => {
  const { user } = req;
  const { connectionId, isApproved } = req.body;

  if (connectionId === undefined || isApproved === undefined) {
    return res.status(400).send();
  }

  if (!mongoose.Types.ObjectId.isValid(connectionId)) {
    return res.status(400).send();
  }
  (async () => {
    const connection = await Connection.findById(connectionId);
    if (connection.from.equals(user._id)) {
      connection.approvedByFrom = isApproved;
    } else {
      connection.approvedByTo = isApproved;
    }
    return res.send(await connection.save());
  })().catch(next);
});

// - post ----
route.get("/posts", authCheck, (req, res, next) => {
  const { user } = req;
  const { isActive } = req.query;
  (async () => {
    const query = {
      by: user
    };
    if (isActive) {
      query.isActive = isActive;
    }
    const posts = await Post.find(query);

    res.send(posts);
  })().catch(next);
});

route.post("/createPost", authCheck, (req, res, next) => {
  const { user } = req;
  const { title, description } = req.body;
  (async () => {
    const post = new Post({
      by: user,
      title: title,
      description: description
    });
    await post.save();
    res.send(post);
  })().catch(next);
});

route.post("/updatePost", authCheck, (req, res, next) => {
  const { user } = req;
  const { postID, title, description } = req.body;

  (async () => {
    if (!postID) {
      return res.status(400).send();
    }

    if (!mongoose.Types.ObjectId.isValid(postID)) {
      return res.status(400).send();
    }

    const post = await Post.findById(postID);
    if (!post) {
      return res.status(400).send();
    }

    if (post.by.equal(user._id)) {
      return res.status(401).send();
    }

    Object.assign(post, { title, description });
    await post.save();
    res.send(post);
  })().catch(next);
});

// - shareLog
route.get("/shareLogs", authCheck, (req, res, next) => {
  const { isIn, isOut } = req.query;
  const { user } = req;

  if (!isIn && !isOut) {
    return res.status(400).send();
  }
  const query = {};
  if (isIn) {
    query.isIn = isIn;
  }
  if (isOut) {
    query.isOut = isOut;
  }

  (async () => {
    const shareLogs = await ShareLog.find(query);
    res.send(shareLogs);
  })().catch(next);
});

route.post("/createShareLog", authCheck, (req, res, next) => {
  const { user } = req;
  const { postID } = req.body;

  if (!mongoose.Types.ObjectId.isValid(postID)) {
    return res.status(400).send();
  }

  (async () => {
    const shareLog = new ShareLog({
      post: postID,
      borrower: user.id
    });
    await post.shareLog().save();
    res.send(shareLog);
  })().catch(next);
});

route.post("/updateInShareLog", authCheck, (req, res, next) => {
  const { user } = req;
  const { shareLogID, isReturning } = req.body;

  if (!mongoose.Types.ObjectId.isValid(shareLogID)) {
    return res.status(400).send();
  }

  (async () => {
    const shareLog = await ShareLog.findById(shareLogID);
    if (!shareLog) {
      return res.status(400).send();
    }

    if (!shareLog.borrower.equal(user._id)) {
      return res.status(401).send();
    }

    shareLog.isReturning = isReturning;
    await shareLog.save();
    res.send(shareLog);
  })().catch(next);
});

route.post("/updateOutShareLog", authCheck, (req, res, next) => {
  const { user } = req;
  const { shareLogID, isReturned } = req.body;

  if (!mongoose.Types.ObjectId.isValid(shareLogID)) {
    return res.status(400).send();
  }

  (async () => {
    const shareLog = await ShareLog.findById(shareLogID).populate("post");
    if (!shareLog) {
      return res.status(400).send();
    }

    if (!shareLog.post.by.equal(user._id)) {
      return res.status(401).send();
    }

    shareLog.isReturned = isReturned;
    await shareLog.save();
    res.send(shareLog);
  })().catch(next);
});

module.exports = route;
