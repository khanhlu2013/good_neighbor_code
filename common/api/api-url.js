import RouteNode from "route-node";

const NODE_ENV = get_NODE_ENV();
let BACKEND_BASE_URL;
let BACKEND_PORT;

if (NODE_ENV === "production") {
  BACKEND_BASE_URL = "https://goodneighbor-backend-test.herokuapp.com";
  BACKEND_PORT = "";
} else {
  BACKEND_BASE_URL = "http://localhost";
  //BACKEND_BASE_URL = "http://10.0.2.2"; //this is for android emulator.since adroid emulator has it own localhost or 127.0.0.1 ; 10.0.2.2 is a alias for local host for android emulator for development purpose.
  BACKEND_PORT = "3001";
}

const BACKEND_URL = `${BACKEND_BASE_URL}:${BACKEND_PORT}`;

const API_NODE = new RouteNode("", "", [
  new RouteNode("profile", "/profile", [
    { name: "authCheck", path: "/authCheck" },
    { name: "logout", path: "/logout" },
    { name: "searchEmail", path: "/searchEmail?:email" },
    // - connection
    { name: "connections", path: "/connections" },
    { name: "createConnection", path: "/createConnection" },
    { name: "updateConnection", path: "/updateConnection" },
    // - post
    { name: "createPost", path: "/createPost" },
    { name: "updatePost", path: "/updatePost" },
    { name: "awareReturnPost", path: "/awareReturnPost" },
    // - outPosts
    { name: "outPosts", path: "/outPosts" },
    // - inPosts
    { name: "inPosts", path: "/inPosts" },
    // - share
    { name: "share", path: "/share" },
    { name: "createShare", path: "/createShare" },
    { name: "deleteShare", path: "/deleteShare" },
    { name: "approveShare", path: "/approveShare" },
    { name: "returnShare", path: "/returnShare" },
    { name: "awareApproveShare", path: "/awareApproveShare" }
  ]),
  new RouteNode("auth", "/auth", [
    { name: "google", path: "/google" },
    {
      name: "backdoorLogin",
      path: "/backdoorLogin?:email?:name"
    }
  ])
]);

const API_URL = (dottedPath, params) => {
  const path = API_NODE.buildPath(dottedPath, params);
  if (!path) {
    throw Error(`${dottedPath} is not found!`);
  }

  return BACKEND_URL + path;
};

export default API_URL;

// - private ---
function get_NODE_ENV() {
  /*
    NOTE that this code is in common project. 
    
    When is is build together with the web app by CreateReactApp, process.env.NODE_ENV will be set based on npm run or npm test or npm build which set it to "development" or "test" or "production"

    When it is build by expo mobile/reactNative app, it will have to be set by expo. i am not sure what is this going to be but i know for sure that for local development, this var is not "production" . Since i haven't build production code for mobile, lets just move on here. 

  */
  return process.env.NODE_ENV;
}
