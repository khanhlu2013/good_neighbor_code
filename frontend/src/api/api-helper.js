import { Connection } from "../model/connection";
import { User } from "../model/user";
import { Post } from "../model/post";
import { Share } from "../model/share";

const rawToConnection = raw => {
  const {
    _id: id,
    from: { _id: fromID, email: fromEmail, name: fromName },
    to: { _id: toID, email: toEmail, name: toName },
    approvedByTo,
    approvedByFrom
  } = raw;

  const from = new User(fromID, fromEmail, fromName);
  const to = new User(toID, toEmail, toName);
  return new Connection(id, from, to, approvedByTo, approvedByFrom);
};

const rawsToConnections = raws => {
  return raws.map(raw => rawToConnection(raw));
};

function rawsToPosts(raws) {
  return raws.map(raw => {
    const { post: postRaw, user: userRaw } = raw;
    const sharesRaw = raw.shares.filter(share => share._id !== undefined);

    const shares = sharesRaw.map(shareRaw => {
      const { _id: userId, email, name } = shareRaw.borrower;
      const borrower = new User(userId, email, name);
      return new Share(
        shareRaw._id,
        borrower,
        shareRaw.dateCreated,
        shareRaw.isApprovedByFrom,
        shareRaw.isReturnedByTo,
        null //post to be set later
      );
    });
    const { _id: userId, email: userEmail, name: userName } = userRaw;
    const post = new Post(
      postRaw._id,
      new User(userId, userEmail, userName),
      postRaw.isActive,
      postRaw.title,
      postRaw.description,
      postRaw.dateCreated,
      shares
    );
    for (const share of post.shares) {
      share.setPost(post);
    }
    return post;
  });
}

export { rawsToConnections, rawToConnection, rawsToPosts };
