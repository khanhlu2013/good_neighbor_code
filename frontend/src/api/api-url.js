import RouteNode from "route-node";

const { NODE_ENV } = process.env;
let BACKEND_BASE_URL;
let BACKEND_PORT;

if (NODE_ENV === "production") {
  BACKEND_BASE_URL = "https://goodneighbor-backend-test.herokuapp.com";
  BACKEND_PORT = "";
} else {
  BACKEND_BASE_URL = "http://localhost";
  if (process.env.REACT_APP_NODE_ENV === "test") {
    BACKEND_PORT = ":3001";
  } else {
    BACKEND_PORT = ":3000";
  }
}

const BACKEND_URL = `${BACKEND_BASE_URL}${BACKEND_PORT}`;

const API_NODE = new RouteNode("", "", [
  new RouteNode("profile", "/profile", [
    { name: "authCheck", path: "/authCheck" },
    { name: "logout", path: "/logout" },
    { name: "searchEmail", path: "/searchEmail?:email" },
    // - connection
    { name: "connections", path: "/connections" },
    { name: "createConnection", path: "/createConnection" },
    { name: "updateConnection", path: "/updateConnection" },
    // - post
    { name: "createPost", path: "/createPost" },
    { name: "updatePost", path: "/updatePost" },
    { name: "awareReturnPost", path: "/awareReturnPost" },
    // - outPosts
    { name: "outPosts", path: "/outPosts" },
    // - inPosts
    { name: "inPosts", path: "/inPosts" },
    // - share
    { name: "share", path: "/share" },
    { name: "createShare", path: "/createShare" },
    { name: "deleteShare", path: "/deleteShare" },
    { name: "approveShare", path: "/approveShare" },
    { name: "returnShare", path: "/returnShare" },
    { name: "awareApproveShare", path: "/awareApproveShare" }
  ]),
  new RouteNode("auth", "/auth", [
    { name: "google", path: "/google" },
    {
      name: "backdoorLogin",
      path: "/backdoorLogin?:email?:name"
    }
  ])
]);

const API_URL = (dottedPath, params) => {
  const path = API_NODE.buildPath(dottedPath, params);
  if (!path) {
    throw Error(`${dottedPath} is not found!`);
  }

  return BACKEND_URL + path;
};

export { API_URL };
