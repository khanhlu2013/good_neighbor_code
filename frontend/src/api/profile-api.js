import keys from "../configs/keys.js";

const profile = async () => {
  const request = await fetch(keys.API_URL("profile"), {
    credentials: "include"
  });
  if (request.status === 401) {
    return null;
  }
  return await request.json();
};

const searchEmail = email => {
  return get("profile.searchEmail", { email });
};

const connections = () => {
  return get("profile.connections", {});
};

const modifyConnection = (connectionId, isApproved) => {
  return post("profile.modifyConnection", {
    connectionId,
    isApproved
  });
};

const createConnection = userIdToAdd => {
  return post("profile.createConnection", {
    userIdToAdd
  });
};

const posts = async () => {
  return get("profile.posts", {});
};

const crudPost = async (postID, title, description) => {
  const crudedPost = await post("profile.crudPost", {
    postID,
    title,
    description
  });
  return crudedPost;
};

//- helper ----

async function getJSON(response) {
  const text = await response.text();
  if (!text) {
    return null;
  }
  return JSON.parse(text);
}

async function post(dottedPath, params) {
  const response = await fetch(keys.API_URL(dottedPath), {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(params),
    credentials: "include"
  });

  return await getJSON(response);
}

async function get(dottedPath, params) {
  const url = keys.API_URL(dottedPath, params);
  const response = await fetch(url, { credentials: "include" });
  return await getJSON(response);
}

//-------------

const API = {
  profile,
  searchEmail,
  connections,
  modifyConnection,
  createConnection,
  posts,
  crudPost
};
export { API };
