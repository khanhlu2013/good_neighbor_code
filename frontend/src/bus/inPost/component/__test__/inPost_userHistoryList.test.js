import React from "react";
import { shallow } from "enzyme";

import InPostUserHistoryList from "../inPost_userHistoryList";
import { rawsToPosts } from "../../../../api/_private_api_helper";
import PostListNoDataWebView from "../../../post/view/postListNoData.webView";

describe("inPost_userHistoryList", () => {
  it("render empty data correctly", () => {
    const wrap = shallow(<InPostUserHistoryList shares={[]} />);
    expect(wrap).toMatchSnapshot();
    expect(wrap.find(PostListNoDataWebView)).toHaveLength(1);
  });

  it("to sort data correctly", () => {
    const before = "2018-11-11T19:07:52.731Z";
    const after = "2018-11-11T19:07:54.392Z";
    let returnToFriendDate;
    let returnToStrangerDate;
    let shares;
    let wrap;

    //friend - before
    returnToFriendDate = before;
    returnToStrangerDate = after;
    shares = create_me_returnBorrowFrom_friendAndStranger_fixture(
      returnToFriendDate,
      returnToStrangerDate
    );
    wrap = shallow(<InPostUserHistoryList shares={shares} />);
    expect(wrap).toMatchSnapshot("friend return before stranger");

    //friend - after
    returnToFriendDate = after;
    returnToStrangerDate = before;
    shares = create_me_returnBorrowFrom_friendAndStranger_fixture(
      returnToFriendDate,
      returnToStrangerDate
    );
    wrap = shallow(<InPostUserHistoryList shares={shares} />);
    expect(wrap).toMatchSnapshot("friend return after stranger");
  });
});

function create_me_returnBorrowFrom_friendAndStranger_fixture(
  friendReturnDateStr,
  strangerReturnDateStr
) {
  const rawFixture = [
    {
      shares: [
        {
          _id: "5be87de80c0c37087779f79e",
          isAwareApprove: true,
          isReturn: true,
          isAwareReturn: false,
          post: "5be87ddf0c0c37087779f79d",
          borrower: {
            _id: "5be87d7a3ade200000fc9539",
            email: "me@me.com",
            name: "I Myself Me"
          },
          dateCreate: "2018-11-11T19:07:20.808Z",
          __v: 0,
          isApprove: true,
          dateReturn: strangerReturnDateStr
        }
      ],
      user: {
        _id: "5be87d7a3ade200000fc9538",
        email: "stranger@person.com",
        name: "Stranger Person"
      },
      post: {
        _id: "5be87ddf0c0c37087779f79d",
        isActive: true,
        user: "5be87d7a3ade200000fc9538",
        title: "stranger",
        description: "stranger",
        dateCreate: "2018-11-11T19:07:11.934Z",
        __v: 0
      }
    },
    {
      shares: [
        {
          _id: "5be87dea0c0c37087779f79f",
          isAwareApprove: true,
          isReturn: true,
          isAwareReturn: false,
          post: "5be87dcf0c0c37087779f79c",
          borrower: {
            _id: "5be87d7a3ade200000fc9539",
            email: "me@me.com",
            name: "I Myself Me"
          },
          dateCreate: "2018-11-11T19:07:22.385Z",
          __v: 0,
          isApprove: true,
          dateReturn: friendReturnDateStr
        }
      ],
      user: {
        _id: "5be87d7a3ade200000fc9537",
        email: "my@friend.com",
        name: "My Friend"
      },
      post: {
        _id: "5be87dcf0c0c37087779f79c",
        isActive: true,
        user: "5be87d7a3ade200000fc9537",
        title: "friend",
        description: "friend",
        dateCreate: "2018-11-11T19:06:55.497Z",
        __v: 0
      }
    }
  ];
  const posts = rawsToPosts(rawFixture);
  const [post1, post2] = posts;
  const [share1] = post1.shares;
  const [share2] = post2.shares;
  return [share1, share2];
}
