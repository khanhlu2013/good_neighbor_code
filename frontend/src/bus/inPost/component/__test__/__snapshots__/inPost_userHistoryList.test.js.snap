// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inPost_userHistoryList render empty data correctly 1`] = `<PostListNoData />`;

exports[`inPost_userHistoryList to sort data correctly: friend return after stranger 1`] = `
<table
  className="table table-sm table-striped table-bordered"
  id="InShareReturnTable-react"
>
  <thead
    className="thead-light"
  >
    <tr>
      <th>
        date
      </th>
      <th>
        post
      </th>
      <th>
        lender
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="5be87dea0c0c37087779f79f"
    >
      <td>
        xx/xx/xx
      </td>
      <td>
        friend
      </td>
      <td>
        My Friend, my@friend.com
      </td>
    </tr>
    <tr
      key="5be87de80c0c37087779f79e"
    >
      <td>
        xx/xx/xx
      </td>
      <td>
        stranger
      </td>
      <td>
        Stranger Person, stranger@person.com
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`inPost_userHistoryList to sort data correctly: friend return before stranger 1`] = `
<table
  className="table table-sm table-striped table-bordered"
  id="InShareReturnTable-react"
>
  <thead
    className="thead-light"
  >
    <tr>
      <th>
        date
      </th>
      <th>
        post
      </th>
      <th>
        lender
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="5be87de80c0c37087779f79e"
    >
      <td>
        xx/xx/xx
      </td>
      <td>
        stranger
      </td>
      <td>
        Stranger Person, stranger@person.com
      </td>
    </tr>
    <tr
      key="5be87dea0c0c37087779f79f"
    >
      <td>
        xx/xx/xx
      </td>
      <td>
        friend
      </td>
      <td>
        My Friend, my@friend.com
      </td>
    </tr>
  </tbody>
</table>
`;
